fortanix-vme/nitro-attestation-verify/src/lib.rs

```diff
@@ -0,0 +1,425 @@
+#[macro_use]
+extern crate lazy_static;
+
+pub use aws_nitro_enclaves_cose::crypto::{Hash, SigningPublicKey};
+use aws_nitro_enclaves_cose::CoseSign1;
+use serde::Deserialize;
+use serde_bytes::ByteBuf;
+use std::collections::BTreeMap;
+use std::marker::PhantomData;
+
+#[cfg(feature = "mbedtls")]
+use ::mbedtls::{alloc::List as MbedtlsList, x509::{Certificate, VerifyError}};
+
+#[cfg(feature = "mbedtls")]
+mod mbedtls;
+
+#[cfg(feature = "mbedtls")]
+use crate::mbedtls::Mbedtls;
+
+pub trait VerificationType {}
+
+#[derive(Clone, Debug)]
+pub struct Verified;
+impl VerificationType for Verified {}
+
+#[derive(Clone, Debug)]
+pub struct Unverified;
+impl VerificationType for Unverified {}
+
+#[derive(Clone, Debug)]
+pub struct Expired;
+impl VerificationType for Expired {}
+
+pub trait SafeToDeserialize {}
+impl SafeToDeserialize for Unverified {}
+
+#[derive(Debug, Deserialize)]
+#[serde(bound(deserialize = "V: SafeToDeserialize"))]
+pub struct AttestationDocument<V: VerificationType = Verified> {
+    module_id: String,
+    timestamp: u64,
+    digest: String,
+    pcrs: BTreeMap<u32, ByteBuf>,
+    certificate: ByteBuf,
+    cabundle: Vec<ByteBuf>,
+    public_key: Option<ByteBuf>,
+    user_data: Option<ByteBuf>,
+    nonce: Option<ByteBuf>,
+    #[serde(skip)]
+    cose: Option<CoseSign1>,
+    #[serde(skip)]
+    type_: PhantomData<V>,
+}
+
+impl<V: VerificationType> AttestationDocument<V> {
+    pub fn module_id(&self) -> &String {
+        &self.module_id
+    }
+
+    pub fn timestamp(&self) -> u64 {
+        self.timestamp
+    }
+
+    pub fn digest(&self) -> &String {
+        &self.digest
+    }
+
+    pub fn pcrs(&self) -> &BTreeMap<u32, ByteBuf> {
+        &self.pcrs
+    }
+
+    pub fn certificate(&self) -> &ByteBuf {
+        &self.certificate
+    }
+
+    pub fn cabundle(&self) -> &Vec<ByteBuf> {
+        &self.cabundle
+    }
+
+    pub fn public_key(&self) -> Option<&ByteBuf> {
+        self.public_key.as_ref()
+    }
+
+    pub fn user_data(&self) -> Option<&ByteBuf> {
+        self.user_data.as_ref()
+    }
+
+    pub fn nonce(&self) -> Option<&ByteBuf> {
+        self.nonce.as_ref()
+    }
+}
+
+impl AttestationDocument<Unverified> {
+    pub fn from_slice<H: Hash>(data: &[u8]) -> Result<Self, NitroError> {
+        // https://docs.aws.amazon.com/enclaves/latest/user/verify-root.html
+        let cose = CoseSign1::from_bytes(data)
+            .map_err(|err| NitroError::CoseParsingFailure(format!("cose error: {:?}", err)))?;
+        let payload = cose
+            .get_payload::<H>(None)
+            .map_err(|err| NitroError::CoseParsingFailure(format!("cose error: {:?}", err)))?;
+        let mut doc: AttestationDocument<Unverified> = serde_cbor::from_slice(&payload)
+            .map_err(|e| NitroError::PayloadParsingFailure(format!("payload error: {}", e)))?;
+        doc.cose = Some(cose);
+        Ok(doc)
+    }
+
+    /// Verifies the `AttestationDocument` given `root_certs` as DER-encoded trusted root CAs
```

Comment:
Add warning that even an `AttestationDocument<Verified>` may be a debug enclave

Reply:
Debug enclave can be checked by looking if pcr0 is all 0s

Reply:
that can be done here as well since you do have the pcrs or can be done in the caller

Reply:
Do you happen to remember where you read that? I can only find the [aws-cli run-enclave](https://docs.aws.amazon.com/ko_kr/enclaves/latest/user/cmd-nitro-run-enclave.html) documentation, but that hints that every PCR is filled with 0s, which isn't the case.

Reply:
https://docs.aws.amazon.com/enclaves/latest/user/enclaves-user.pdf

```
Note
Enclaves booted in debug mode generate attestation documents with PCRs that are made
up entirely of zeros (000000000000000000000000000000000000000000000000). These
attestation documents cannot be used for cryptographic attestation.
```

Reply:
Thanks! But that's still an odd location for that documentation. Will add some comment soon

Reply:
Added documentation